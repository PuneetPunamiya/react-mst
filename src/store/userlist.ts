import { types, Instance } from "mobx-state-tree";
import { values } from "mobx";

export const User = types.model({
  name: types.optional(types.string, ""),
  age: types.optional(types.integer, 0),
  class: types.optional(types.string, ""),
});

export const Store = types
  .model({
    users: types.array(User),
    filter: types.optional(types.array(types.string), []),
    sort: types.optional(types.string, ""),
  })
  .views((self) => ({
    get count() {
      return self.users.length;
    },
    get userList() {
      return filterAndSortUsers(values(self.users), self.filter, self.sort);
    },
  }))
  .actions((self) => ({
    add(item: UserItem) {
      self.users.push(item);
    },
    setSort(sort: string) {
      self.sort = sort;
    },
    setFilter(filter: any) {
      self.filter = filter;
    },
  }));

export interface UserItem extends Instance<typeof User> {}

function filterAndSortUsers(users: any, filter: any, sort: any) {
  if (filter.length === 0 && sort === "") {
    return users;
  }
  if (filter.length === 0) {
    if (sort === "name") {
      return users.sort((a: UserItem, b: UserItem) =>
        a.name > b.name ? 1 : a.name === b.name ? 0 : -1
      );
    } else if (sort === "age") {
      return users.sort((a: UserItem, b: UserItem) =>
        a.age > b.age ? 1 : a.age === b.age ? 0 : -1
      );
    } else {
      return users;
    }
  }
  if (sort === "name") {
    return users
      .filter((u: UserItem) => filter.includes(u.class))
      .sort((a: UserItem, b: UserItem) =>
        a.name > b.name ? 1 : a.name === b.name ? 0 : -1
      );
  } else if (sort === "age") {
    return users
      .filter((u: UserItem) => filter.includes(u.class))
      .sort((a: UserItem, b: UserItem) =>
        a.age > b.age ? 1 : a.age === b.age ? 0 : -1
      );
  } else {
    return users.filter((u: UserItem) => filter.includes(u.class));
  }
}
