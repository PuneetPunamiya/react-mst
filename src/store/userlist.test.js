import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree";
import { User, Store } from "./userlist";
import * as fs from "fs";
import { when } from "mobx";

const userFetcher = () =>
  Promise.resolve(JSON.parse(fs.readFileSync("./src/store/users.json")));

it("can create an user", () => {
  const user = User.create({
    id: 1,
    name: "test",
    age: 12,
    class: "SY",
  });

  expect(user.name).toBe("test");
  expect(user.age).toBe(12);
  expect(user.class).toBe("SY");
});

it("can create a store", (done) => {
  const store = Store.create({}, { fetch: userFetcher });
  expect(store.count).toBe(0);
  expect(store.isLoading).toBe(true);
  when(
    () => !store.isLoading,
    () => {
      expect(store.count).toBe(6);
      expect(store.isLoading).toBe(false);
      done();
    }
  );
});

it("can add user in a store", (done) => {
  const store = Store.create({}, { fetch: userFetcher });
  expect(store.count).toBe(0);
  expect(store.isLoading).toBe(true);
  when(
    () => !store.isLoading,
    () => {
      expect(store.isLoading).toBe(false);

      const states = [];
      onSnapshot(store, (s) => states.push(s));
      const patches = [];
      onPatch(store, (p) => patches.push(p));

      store.add({ id: 43, name: "test", age: 43, class: "SY" });
      expect(store.count).toBe(7);
      expect(getSnapshot(store)).toMatchSnapshot();
      expect(patches).toMatchSnapshot();
      done();
    }
  );
});

it("can get users sorted", (done) => {
  const store = Store.create({}, { fetch: userFetcher });
  expect(store.count).toBe(0);
  expect(store.isLoading).toBe(true);
  when(
    () => !store.isLoading,
    () => {
      expect(store.isLoading).toBe(false);

      const states = [];
      onSnapshot(store, (s) => states.push(s));
      const patches = [];
      onPatch(store, (p) => patches.push(p));

      //Sorted by name
      store.setSort("name");
      const list = store.userList;
      expect(list[0].name).toBe("chandler");
      expect(getSnapshot(store)).toMatchSnapshot();
      expect(patches).toMatchSnapshot();
      done();
    }
  );
});

it("can get users filtered by class", (done) => {
  const store = Store.create({}, { fetch: userFetcher });
  expect(store.count).toBe(0);
  expect(store.isLoading).toBe(true);
  when(
    () => !store.isLoading,
    () => {
      expect(store.isLoading).toBe(false);

      const states = [];
      onSnapshot(store, (s) => states.push(s));
      const patches = [];
      onPatch(store, (p) => patches.push(p));

      store.setFilter(["FY"]);
      const list = store.userList;
      expect(list.length).toBe(3);
      expect(getSnapshot(store)).toMatchSnapshot();
      expect(patches).toMatchSnapshot();
      done();
    }
  );
});

it("can get users filtered by class and sorted", (done) => {
  const store = Store.create({}, { fetch: userFetcher });
  expect(store.count).toBe(0);
  expect(store.isLoading).toBe(true);
  when(
    () => !store.isLoading,
    () => {
      expect(store.isLoading).toBe(false);

      const states = [];
      onSnapshot(store, (s) => states.push(s));
      const patches = [];
      onPatch(store, (p) => patches.push(p));

      store.setSort("name");
      store.setFilter(["FY"]);
      const list = store.userList;
      expect(list.length).toBe(3);
      expect(list[0].name).toBe("joey");
      expect(getSnapshot(store)).toMatchSnapshot();
      expect(patches).toMatchSnapshot();
      done();
    }
  );
});
