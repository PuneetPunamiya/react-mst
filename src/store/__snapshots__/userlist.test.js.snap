// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add user in a store 1`] = `
Object {
  "filter": Array [],
  "isLoading": false,
  "sort": "",
  "users": Array [
    Object {
      "age": 18,
      "class": "FY",
      "id": 11,
      "name": "rachel",
    },
    Object {
      "age": 21,
      "class": "SY",
      "id": 12,
      "name": "monica",
    },
    Object {
      "age": 27,
      "class": "FY",
      "id": 13,
      "name": "joey",
    },
    Object {
      "age": 22,
      "class": "TY",
      "id": 14,
      "name": "phoebe",
    },
    Object {
      "age": 17,
      "class": "FY",
      "id": 15,
      "name": "ross",
    },
    Object {
      "age": 28,
      "class": "SY",
      "id": 16,
      "name": "chandler",
    },
    Object {
      "age": 43,
      "class": "SY",
      "id": 43,
      "name": "test",
    },
  ],
}
`;

exports[`can add user in a store 2`] = `
Array [
  Object {
    "op": "add",
    "path": "/users/6",
    "value": Object {
      "age": 43,
      "class": "SY",
      "id": 43,
      "name": "test",
    },
  },
]
`;

exports[`can get users filtered by class 1`] = `
Object {
  "filter": Array [
    "FY",
  ],
  "isLoading": false,
  "sort": "",
  "users": Array [
    Object {
      "age": 18,
      "class": "FY",
      "id": 11,
      "name": "rachel",
    },
    Object {
      "age": 21,
      "class": "SY",
      "id": 12,
      "name": "monica",
    },
    Object {
      "age": 27,
      "class": "FY",
      "id": 13,
      "name": "joey",
    },
    Object {
      "age": 22,
      "class": "TY",
      "id": 14,
      "name": "phoebe",
    },
    Object {
      "age": 17,
      "class": "FY",
      "id": 15,
      "name": "ross",
    },
    Object {
      "age": 28,
      "class": "SY",
      "id": 16,
      "name": "chandler",
    },
  ],
}
`;

exports[`can get users filtered by class 2`] = `
Array [
  Object {
    "op": "add",
    "path": "/filter/0",
    "value": "FY",
  },
]
`;

exports[`can get users filtered by class and sorted 1`] = `
Object {
  "filter": Array [
    "FY",
  ],
  "isLoading": false,
  "sort": "name",
  "users": Array [
    Object {
      "age": 18,
      "class": "FY",
      "id": 11,
      "name": "rachel",
    },
    Object {
      "age": 21,
      "class": "SY",
      "id": 12,
      "name": "monica",
    },
    Object {
      "age": 27,
      "class": "FY",
      "id": 13,
      "name": "joey",
    },
    Object {
      "age": 22,
      "class": "TY",
      "id": 14,
      "name": "phoebe",
    },
    Object {
      "age": 17,
      "class": "FY",
      "id": 15,
      "name": "ross",
    },
    Object {
      "age": 28,
      "class": "SY",
      "id": 16,
      "name": "chandler",
    },
  ],
}
`;

exports[`can get users filtered by class and sorted 2`] = `
Array [
  Object {
    "op": "replace",
    "path": "/sort",
    "value": "name",
  },
  Object {
    "op": "add",
    "path": "/filter/0",
    "value": "FY",
  },
]
`;

exports[`can get users sorted 1`] = `
Object {
  "filter": Array [],
  "isLoading": false,
  "sort": "name",
  "users": Array [
    Object {
      "age": 18,
      "class": "FY",
      "id": 11,
      "name": "rachel",
    },
    Object {
      "age": 21,
      "class": "SY",
      "id": 12,
      "name": "monica",
    },
    Object {
      "age": 27,
      "class": "FY",
      "id": 13,
      "name": "joey",
    },
    Object {
      "age": 22,
      "class": "TY",
      "id": 14,
      "name": "phoebe",
    },
    Object {
      "age": 17,
      "class": "FY",
      "id": 15,
      "name": "ross",
    },
    Object {
      "age": 28,
      "class": "SY",
      "id": 16,
      "name": "chandler",
    },
  ],
}
`;

exports[`can get users sorted 2`] = `
Array [
  Object {
    "op": "replace",
    "path": "/sort",
    "value": "name",
  },
]
`;
